eval/adjust.rkt

```diff
@@ -66,53 +65,60 @@
-  (vector-copy! vprecs 0 vprecs-new)
-
-  ; Step 5. If precisions have not changed but the point didn't converge. A problem exists - add slack to every op
-  (unless any-false?
-    (set-rival-machine-bumps! machine (add1 bumps))
-    (define slack (get-slack))
-    (for ([prec (in-vector vprecs)]
-          [n (in-range (vector-length vprecs))])
-      (define prec* (min (*rival-max-precision*) (+ prec slack)))
-      (when (equal? prec* (*rival-max-precision*))
-        (*sampling-iteration* (*rival-max-iterations*)))
-      (vector-set! vprecs n prec*))
-    (vector-fill! vrepeats #f)))
+  (vector-copy! vprecs 0 vprecs-new))
 
 ; This function goes through ivec and vregs and calculates (+ ampls base-precisions) for each operator in ivec
-; Roughly speaking:
-;   vprecs-new[i] = min( *rival-max-precision* max( *base-tuning-precision* (+ intro vstart-precs[i])),
-;   intro = get-ampls(parent)
-(define (precision-tuning ivec vregs vprecs-new varc vstart-precs)
-  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)] ; reversed over ivec
-        [n (in-range (- (vector-length vregs) 1) -1 -1)]) ; reversed over indices of vregs
-
-    (define op (car instr)) ; current operation
+; Roughly speaking, the upper precision bound is calculated as:
+;   vprecs-max[i] = (+ max-prec vstart-precs[i]), where min-prec < (+ max-prec vstart-precs[i]) < max-prec
+;   max-prec = (car (get-bounds parent))
+(define (precision-tuning ivec vregs vprecs-max varc vstart-precs vuseful)
+  (define vprecs-min (make-vector (vector-length ivec) 0))
+  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)]
+        [useful? (in-vector vuseful (- (vector-length vuseful) 1) -1 -1)]
+        [n (in-range (- (vector-length vregs) 1) -1 -1)]
+        #:when useful?)
+    (define op (car instr))
     (define tail-registers (cdr instr))
-    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
-    (define output (vector-ref vregs n)) ; output of the current instr
-
-    (define intro (vector-ref vprecs-new (- n varc))) ; intro for the current instruction
-    (define ampls (get-ampls op output srcs)) ; ampls for the tail instructions
-
-    (define final-parent-precision
-      (max (+ intro (vector-ref vstart-precs (- n varc))) (*base-tuning-precision*)))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define max-prec (vector-ref vprecs-max (- n varc))) ; upper precision bound given from parent
+    (define min-prec (vector-ref vprecs-min (- n varc))) ; lower precision bound given from parent
+
+    ; Final precision assignment based on the upper bound
+    (define final-precision
+      (min (max (+ max-prec (vector-ref vstart-precs (- n varc))) (*rival-min-precision*))
+           (*rival-max-precision*)))
+    (vector-set! vprecs-max (- n varc) final-precision)
+
+    ; Early stopping
+    (match (*lower-bound-early-stopping*)
```

Comment:
Yes

Reply:
I guess I can delete this `match` and `if` statements everywhere since it is will be _on_ by default, but maybe later for rerunning our latest experiments it is good to have it.

Reply:
Yeah let's keep it.